src/NexusMods.App.UI/Pages/LibraryPage/LibraryViewModel.cs

```diff
@@ -283,42 +238,50 @@ await Parallel.ForAsync(
             cancellationToken: cancellationToken
         );
     }
+}
 
-    private static (ITreeDataGridSource<LibraryItemModel> source, Observable<LibraryItemModel[]> selectionObservable) CreateSource(IEnumerable<LibraryItemModel> models, bool createHierarchicalSource)
+public class LibraryTreeDataGridAdapter : TreeDataGridAdapter<LibraryItemModel>
+{
+    private readonly ILibraryDataProvider[] _libraryDataProviders;
+    private readonly ConnectableObservable<DateTime> _ticker;
+
+    public LibraryTreeDataGridAdapter(IServiceProvider serviceProvider, ConnectableObservable<DateTime> ticker)
     {
-        if (createHierarchicalSource)
-        {
-            var source = new HierarchicalTreeDataGridSource<LibraryItemModel>(models);
-            var selectionObservable = Observable.FromEventHandler<TreeSelectionModelSelectionChangedEventArgs<LibraryItemModel>>(
-                addHandler: handler => source.RowSelection!.SelectionChanged += handler,
-                removeHandler: handler => source.RowSelection!.SelectionChanged -= handler
-            ).Select(tuple => tuple.e.SelectedItems.NotNull().ToArray());
-
-            AddColumns(source.Columns, viewAsTree: true);
-            return (source, selectionObservable);
-        }
-        else
-        {
-            var source = new FlatTreeDataGridSource<LibraryItemModel>(models);
-            var selectionObservable = Observable.FromEventHandler<TreeSelectionModelSelectionChangedEventArgs<LibraryItemModel>>(
-                addHandler: handler => source.RowSelection!.SelectionChanged += handler,
-                removeHandler: handler => source.RowSelection!.SelectionChanged -= handler
-            ).Select(tuple => tuple.e.SelectedItems.NotNull().ToArray());
-
-            AddColumns(source.Columns, viewAsTree: false);
-            return (source, selectionObservable);
-        }
+        _libraryDataProviders = serviceProvider.GetServices<ILibraryDataProvider>().ToArray();
+        _ticker = ticker;
+    }
+
+    protected override void BeforeModelActivationHook(LibraryItemModel model)
+    {
+        model.Ticker = _ticker;
+    }
+
+    protected override void BeforeModelDeactivationHook(LibraryItemModel model)
+    {
+        model.Ticker = null;
+    }
```

Comment:
I know parent base methods are empty, but good practice would be to call the base methods, right? In case those are updated to do something in the future.